File: original_JAVA/OneLine.java
/*  Beach Fill Calculation
*/

import java.awt.*;
import java.applet.*;
import java.lang.*;
import java.util.Vector;
import ErrorFunction;

public class OneLine extends Applet 
 {
   InputOLPanel panel1;
   PlotOLPanel graph1;
   
   public void init()
  
     {
       setLayout(new BorderLayout());
       Label label1 = new Label(" ");
       add("North",label1);  //spacer
       graph1 = new PlotOLPanel();
       add("Center",graph1);

       panel1 = new InputOLPanel(graph1);
       add("West",panel1);
     }
                            
}      //end of Wavemaker

    class  InputOLPanel extends Panel  
     {  
      PlotOLPanel panel;
      TextField Hin, Tin, din;
      Refract ref;
      public static float el,Y, G;
      public static int jstep = 0;
      Button Calculate, Stop;
      public boolean stop = false;
      Label label1;
/*      Choice wmtype;
      boolean wavemaker = true; //means piston
*/
  
      public InputOLPanel(PlotOLPanel panel)
    
       {
        this.panel = panel;
        setLayout(new GridLayout(6,2));
/*        add(new Label("Choose Wavemaker"));
        wmtype = new Choice();
        wmtype.addItem("Piston");
        wmtype.addItem("Flap");
        add(wmtype);
*/
      
        
        
        add(label1 = new Label("Beach Fill Data:"));
        add(new Label(" "));
        add(new Label("Fill Length (m)?"));
        add(Hin = new TextField("1000.0" ));
        add(new Label("Cross-shore Width (m)?"));
        add(Tin = new TextField("100.0" ));
         add(new Label("Shoreline Diffusivity (m^2/s)?"));
        add(din = new TextField("0.001" ));
 
        add(Calculate = new Button("Calculate"));
        add(Stop = new Button("Stop"));
       }

    public Insets insets()
          {
            return new Insets(10,10,10,10);
          }
    public boolean handleEvent(Event e) 
     {
     
      float klin;
      float Cg, Ks, Kr, angle, C ;
    
      ref= new Refract();
      try 
         {
          if(e.id == Event.ACTION_EVENT)
            {
/*             if(  e.target.equals(wmtype))
               if ("Flap".equals (e.arg) )
                   wavemaker= false;
               else
                   wavemaker=true;
 
  */             
            if( e.target instanceof Button ) 
            { if(e.target == Calculate)
               { el = Float.valueOf(Hin.getText()).floatValue();
                 Y = Float.valueOf(Tin.getText()).floatValue();
                 G = Float.valueOf(din.getText()).floatValue();
                 Hin.setEditable(false);
                 Tin.setEditable(false);
                 din.setEditable(false);
                 stop = false;
                 panel.set(stop);
                 panel.do_wave(el,Y ,G);

                }
             else if (e.target == Stop)
               { stop = true;
                 Hin.setEditable(true);
                 Tin.setEditable(true);
                 din.setEditable(true);
                 panel.set(stop);
               }
            }       //end of Buttons
           }  //end ACTION_EVENT
         }
      catch (NumberFormatException e1) 
         {
          System.out.println(e1);
          }

      return false;
     }

} // end InputOLPanel
    class  PlotOLPanel extends Panel implements Runnable
     {
       OneLinePlot makerplot;
       Image wImage;
       Graphics wGraphics;
       public float H, d, L, T;
       boolean first = false,   stop=false;
       Thread thread;
       int jstep;
       
       public PlotOLPanel()
        {
          setBackground(Color.white);
         }

       public void set(boolean stopp) 
        {
         stop=stopp;
        }

       public Dimension minimumSize()
        {
         return new Dimension(300,100);
        }
     
       public void paint(Graphics g)
        {
         if(!first)
           {
            makerplot = new OneLinePlot(bounds().width,bounds().height);

            wImage = createImage(bounds().width, bounds().height);
            wGraphics = wImage.getGraphics();
            first = true;
           }

         wGraphics.clearRect(0,0,bounds().width,bounds().height);
         makerplot.Draw(wGraphics);
         g.drawImage(wImage,0,0,this);
         }
    
        public void update(Graphics g)
         {
          paint(g);
//          System.out.println("doing update in PlotPanel");
         }
        
        public void do_wave(float el, float Y ,float G )
         {
          makerplot.initialize((float) el,(float)Y,(float) G);

          makerplot.do_wave();
//          System.out.println("doing  in PlotPanel");
          thread = new Thread(this);
          thread.start();

         }
      public void run()
     {
      
       while(!stop)
        {
         makerplot.do_wave( );
         repaint();
         try {thread.sleep(150);} 
         catch(Exception e){ }
        }
     }
   
   }//end of class PlotOLPanel

     class OneLinePlot
      {
       public int length, amplitude, totalh;
       int x[];
       public float el,Y,G,L,T;
       int depth[],uc[],ut[],vt[],vc[];
       int y[],xo[];
       float dz ;
       ErrorFunction errorf;
       double dt,k;
       int dep;
       float famplitude;
       int ncount=12;
       String Mstring = " ";
       String Max = " ";
       String time= " ";
       String Half = " ";
        public int temp;
       public int jstep;
       double fact;
    
       float umag[],vmag[];
       int toffset=5,boffset =10;
       int Depth;
       public int pw=5; //half paddlewidth
       
        public OneLinePlot(int width, int height)
        {
          length = width;
          totalh  = height;
          x = new int[length+2];
          uc = new int[ncount];
          ut = new int[ncount];
          vc = new int[ncount];
          vt = new int[ncount];
          y = new int[length+2];
          xo= new int[length+2];
//          double e1,e2,e3;
          umag = new float[ncount];
          vmag = new float[ncount];
          depth = new int[ncount];
          errorf= new ErrorFunction();
   
          for (int i = 0; i<length;i++)
             { 
              x[i]=0;             
             }
         }
        public void initialize( float he, float eel, float de )
         { 
          
            el=he;
            
            double xi;
            Y=eel;
            G=de;
            jstep=0;
             dt=86400.;  //one day in seconds
            famplitude = (float) (totalh-toffset-boffset);
            amplitude = (int) famplitude;
        
            for (int i = 0; i<length;i++)
              {               
               xi=(double) (i*(1.5*el)/length);
               y[i]=i;
               fact=1.0e07;
               xo[i]=toffset+amplitude-
                 (int) (0.5*amplitude*(errorf.erf(fact*(2.*xi/el+1.))-errorf.erf(fact*(2.*xi/el-1.))));
               
             }
         
         }
  
         public void do_wave()
         {  
            double xi;
            int M;
            double part;
            double half;
                           
            time="time = " + jstep + " days";
          
            fact=el/(4.*Math.sqrt(G*jstep*dt));
            for (int i = 0; i<length;i++)
             { 
               xi=(double) (i*(1.5*el)/length);

               x[i]=toffset+amplitude-(int) (0.5*amplitude*(errorf.erf(fact*(2.*xi/el+1.))-errorf.erf(fact*(2.*xi/el-1.))));
               
             }
            // percent fill in project area
             fact=2.*fact;
             part= (1./(Math.sqrt(Math.PI)*fact))*(Math.exp(-fact*fact)-1.);
             M= (int) (100.*(part +errorf.erf(fact)));
             Mstring="percent= "+M;
             Max="max = "+(int)((int)((toffset+amplitude-x[0])*Y*100./amplitude)/100.)+" m";
             half=((int)(Math.pow((0.46*el),2.)*100./(G*86400*365))/100.);
             Half="half-life="+half+" yrs";
             jstep=jstep+1;

          }
         public void Draw(Graphics g)
         {
          int xl,yl ;
          g.setColor(Color.black);
          g.drawLine(0,amplitude+toffset,length,amplitude+toffset); //x axis
          g.drawLine(0,0,0,totalh);                 //z axis
          g.setColor(Color.yellow);   //color beach
          g.fillRect(0,amplitude+toffset,length,amplitude+toffset); 
          //box panel
          xl=length-1;
          yl=totalh-1;
          g.setColor(Color.black);
          g.drawLine(0,yl,xl,yl);
          g.drawLine(xl,0,xl,xl);
          g.drawLine(0,0,xl,0);
 
          g.setColor(Color.magenta);
          x[length]=amplitude+toffset;
          y[length]=length;
          y[length+1]=0;
          x[length+1]=amplitude+toffset;
          g.fillPolygon(y,x,length+2);
          
          xl=length-130;
          yl=yl-90;
          g.setColor(Color.black);
          g.drawString(time,xl,yl);
          yl=yl+20;
          g.drawString(Max,xl,yl);
          yl=yl+20;
          g.drawString(Mstring,xl,yl);
          yl=yl+20;
          g.drawString(Half,xl,yl);
          g.setColor(Color.blue);
          yl=30;
          g.drawString("Ocean",xl,yl);
          g.setColor(Color.white);
          xl=20;
          yl=totalh-30;
          g.drawString("Fill",xl,yl);
          
          g.setColor(Color.black);

           for(int i =1;i<length;i++)
            g.drawLine(y[i-1], xo[i-1],y[i],xo[i]); // original fill contour
         }
         

        public void update(Graphics g)
         {
          Draw(g);  
         }
      }//end of class OneLinePlot

         
          
        

File: OneLine.js
// JavaScript version of the OneLine class and related components

class OneLine {
  constructor() {
      this.init();
  }

  init() {
      // Set up layout and elements
      const container = document.createElement('div');
      container.style.display = 'flex';

      // Spacer
      const spacer = document.createElement('div');
      spacer.textContent = ' ';
      container.appendChild(spacer);

      // Create the graph container
      this.graph1 = new PlotOLPanel();
      container.appendChild(this.graph1.createPlot());

      // Create input panel
      this.panel1 = new InputOLPanel(this.graph1);
      container.appendChild(this.panel1.createPanel());

      // Append to wave-calculator div in the HTML
      document.getElementById('wave-calculator').appendChild(container);
  }
}

// InputOLPanel class
class InputOLPanel {
  constructor(panel) {
      this.panel = panel;
      this.stop = false;
      this.el = 0;
      this.Y = 0;
      this.G = 0;
      this.jstep = 0;
  }

// Updated createPanel() method for InputOLPanel class

createPanel() {
  const panel = document.createElement('div');
  panel.style.display = 'grid';
  panel.style.gridTemplateColumns = 'auto auto'; // Two columns layout, similar to original Java
  panel.style.gap = '10px';
  panel.style.width = '300px';

  // Beach Fill Data label
  const label = document.createElement('h3');
  label.innerText = 'Beach Fill Data:';
  panel.appendChild(label);
  panel.appendChild(document.createElement('div')); // Spacer for alignment

  // Fill Length input
  panel.appendChild(this.createLabel('Fill Length (m)?'));
  this.Hin = this.createInput('1000.0');
  panel.appendChild(this.Hin);

  // Cross-shore Width input
  panel.appendChild(this.createLabel('Cross-shore Width (m)?'));
  this.Tin = this.createInput('100.0');
  panel.appendChild(this.Tin);

  // Shoreline Diffusivity input
  panel.appendChild(this.createLabel('Shoreline Diffusivity (m&sup2;/s)?'));
  this.din = this.createInput('0.001');
  panel.appendChild(this.din);

  // Spacer to align buttons
  panel.appendChild(document.createElement('div')); // Spacer for layout

  // Buttons
  const calculateButton = this.createButton('Calculate', () => this.calculate());
  panel.appendChild(calculateButton);

  const stopButton = this.createButton('Stop', () => this.stopSimulation());
  panel.appendChild(stopButton);

  return panel;
}

  createLabel(text) {
      const label = document.createElement('label');
      label.innerHTML = text;  // For special characters like m²
      return label;
  }

  createInput(defaultValue) {
      const input = document.createElement('input');
      input.type = 'text';
      input.value = defaultValue;
      input.style.width = '100%';
      return input;
  }

  createButton(text, onClick) {
      const button = document.createElement('button');
      button.innerText = text;
      button.onclick = onClick;
      return button;
  }

  calculate() {
      try {
          this.el = parseFloat(this.Hin.value);
          this.Y = parseFloat(this.Tin.value);
          this.G = parseFloat(this.din.value);

          if (isNaN(this.el) || isNaN(this.Y) || isNaN(this.G)) {
              alert('Please enter valid values for all fields.');
              return;
          }

          this.Hin.disabled = true;
          this.Tin.disabled = true;
          this.din.disabled = true;
          this.stop = false;

          this.panel.setStop(false);
          this.panel.doWave(this.el, this.Y, this.G);

      } catch (error) {
          console.error(error);
      }
  }

  stopSimulation() {
      this.stop = true;
      this.Hin.disabled = false;
      this.Tin.disabled = false;
      this.din.disabled = false;
      this.panel.setStop(true);
  }
}


// PlotOLPanel class
class PlotOLPanel {
  constructor() {
      this.first = false;
      this.stop = false;
      this.thread = null;
      this.makerplot = new OneLinePlot(600, 450);
  }

  createPlot() {
      this.canvas = document.createElement('canvas');
      this.canvas.width = 600;
      this.canvas.height = 450;
      this.ctx = this.canvas.getContext('2d');
      return this.canvas;
  }

  setStop(stopp) {
      this.stop = stopp;
      if (this.stop && this.thread) {
          clearInterval(this.thread);
          this.thread = null;
      }
  }

  doWave(el, Y, G) {
      this.makerplot.initialize(el, Y, G);
      this.makerplot.doWave();

      if (this.thread) {
          clearInterval(this.thread);
      }

      this.thread = setInterval(() => {
          if (!this.stop) {
              this.makerplot.doWave();
              this.paint();
          } else {
              clearInterval(this.thread);
          }
      }, 150);
  }

  paint() {
      this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
      this.makerplot.draw(this.ctx);
  }
}

// OneLinePlot class
class OneLinePlot {
  constructor(width, height) {
      this.length = width;
      this.totalh = height;
      this.x = new Array(this.length + 2).fill(0);
      this.y = new Array(this.length + 2).fill(0);
      this.xo = new Array(this.length + 2).fill(0);
      this.errorf = new ErrorFunction();
      this.jstep = 0;
      this.dt = 86400; // One day in seconds
      this.toffset = 5;
      this.boffset = 10;
      this.amplitude = 0;
  }

  initialize(el, Y, G) {
      this.el = el;
      this.Y = Y;
      this.G = G;
      this.amplitude = Math.floor(this.totalh - this.toffset - this.boffset);

      // Initialize the fill shape based on the error function
      for (let i = 0; i < this.length; i++) {
          const xi = (i * 1.5 * this.el) / this.length;
          const fact = 1.0e7;

          this.xo[i] = this.toffset + this.amplitude - 
              0.5 * this.amplitude * 
              (this.errorf.erf(fact * (2 * xi / this.el + 1)) - this.errorf.erf(fact * (2 * xi / this.el - 1)));
          this.y[i] = i;  // Update y values as x-coordinate
      }
  }

  doWave() {
      const fact = this.el / (4 * Math.sqrt(this.G * this.jstep * this.dt));
      for (let i = 0; i < this.length; i++) {
          const xi = (i * 1.5 * this.el) / this.length;

          this.x[i] = this.toffset + this.amplitude -
              0.5 * this.amplitude * 
              (this.errorf.erf(fact * (2 * xi / this.el + 1)) - this.errorf.erf(fact * (2 * xi / this.el - 1)));
      }

      // Update the text-based calculations for display
      const part = (1 / (Math.sqrt(Math.PI) * fact)) * (Math.exp(-fact * fact) - 1);
      const M = Math.floor(100 * (part + this.errorf.erf(fact)));

      this.Max = `Max = ${Math.floor(((this.toffset + this.amplitude - this.x[0]) * this.Y * 100 / this.amplitude) / 100)} m`;
      this.Half = `Half-life = ${(Math.pow(0.46 * this.el, 2) / (this.G * 86400 * 365)).toFixed(2)} yrs`;
      this.Percent = `Percent = ${M}%`;
      this.Time = `Time = ${this.jstep} days`;

      this.jstep += 1;
  }

  draw(g) {
      // Draw the x-axis and z-axis (like in the original Java)
      g.strokeStyle = 'black';
      g.beginPath();
      g.moveTo(0, this.amplitude + this.toffset);
      g.lineTo(this.length, this.amplitude + this.toffset); // x-axis
      g.moveTo(0, 0);
      g.lineTo(0, this.totalh); // z-axis
      g.stroke();

      // Fill the beach and fill areas
      g.fillStyle = 'yellow';  // Beach color
      g.fillRect(0, this.amplitude + this.toffset, this.length, this.amplitude + this.toffset);

      g.fillStyle = 'magenta';  // Fill color
      g.beginPath();
      for (let i = 0; i < this.length; i++) {
          g.lineTo(this.y[i], this.x[i]);
      }
      g.lineTo(this.length, this.amplitude + this.toffset);
      g.lineTo(0, this.amplitude + this.toffset);
      g.closePath();
      g.fill();

      // Draw textual info on the canvas
      g.fillStyle = 'black';
      g.fillText(this.Time, this.length - 130, this.totalh - 90);
      g.fillText(this.Max, this.length - 130, this.totalh - 70);
      g.fillText(this.Percent, this.length - 130, this.totalh - 50);
      g.fillText(this.Half, this.length - 130, this.totalh - 30);
  }
}


// Instantiate the applet
window.onload = function() {
  new OneLine();
};
